File: composer/modules/web/src/plugins/ballerina/model/tree/abstract-tree/array-type-node.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import _ from 'lodash';
import Node from '../node';

class AbstractArrayTypeNode extends Node {


    setDimensions(newValue, silent, title) {
        const oldValue = this.dimensions;
        title = (_.isNil(title)) ? `Modify ${this.kind}` : title;
        this.dimensions = newValue;

        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'modify-node',
                title,
                data: {
                    attributeName: 'dimensions',
                    newValue,
                    oldValue,
                },
            });
        }
    }

    getDimensions() {
        return this.dimensions;
    }



    setElementType(newValue, silent, title) {
        const oldValue = this.elementType;
        title = (_.isNil(title)) ? `Modify ${this.kind}` : title;
        this.elementType = newValue;

        this.elementType.parent = this;

        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'modify-node',
                title,
                data: {
                    attributeName: 'elementType',
                    newValue,
                    oldValue,
                },
            });
        }
    }

    getElementType() {
        return this.elementType;
    }




    isNullable() {
        return this.nullable;
    }

    setNullable(newValue, silent, title) {
        const oldValue = this.nullable;
        title = (_.isNil(title)) ? `Modify ${this.kind}` : title;
        this.nullable = newValue;
        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'modify-node',
                title,
                data: {
                    attributeName: 'nullable',
                    newValue,
                    oldValue,
                },
            });
        }
    }


    isGrouped() {
        return this.grouped;
    }

    setGrouped(newValue, silent, title) {
        const oldValue = this.grouped;
        title = (_.isNil(title)) ? `Modify ${this.kind}` : title;
        this.grouped = newValue;
        if (!silent) {
            this.trigger('tree-modified', {
                origin: this,
                type: 'modify-node',
                title,
                data: {
                    attributeName: 'grouped',
                    newValue,
                    oldValue,
                },
            });
        }
    }

}

export default AbstractArrayTypeNode;

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/types/ArrayTypeNode.java
/*
*  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
*
*  WSO2 Inc. licenses this file to you under the Apache License,
*  Version 2.0 (the "License"); you may not use this file except
*  in compliance with the License.
*  You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing,
*  software distributed under the License is distributed on an
*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
*  KIND, either express or implied.  See the License for the
*  specific language governing permissions and limitations
*  under the License.
*/
package org.ballerinalang.model.tree.types;

import org.wso2.ballerinalang.compiler.tree.expressions.BLangExpression;

/**
 * {@code ArrayTypeNode} represents an array type node in Ballerina
 * array type node can be of multiple dimensions and each dimension can be sealed with a specific size
 * <p>
 * e.g. int[2][][4];
 *
 * @since 0.94
 */
public interface ArrayTypeNode extends ReferenceTypeNode {
    TypeNode getElementType();

    int getDimensions();

    BLangExpression[] getSizes();
}
